[OpenMP] Introduce the initial support for OpenMP kernel language

shiltian opened the pull request:
This patch starts the support for OpenMP kernel language, basically to write
OpenMP target region in SIMT style, similar to kernel languages such as CUDA.
What included in this first patch is the `ompx_bare` clause for `target teams`
directive. When `ompx_bare` exists, globalization is disabled such that local
variables will not be globalized. The runtime init/deinit function calls will
not be emitted. That being said, almost all OpenMP executable directives are
not supported in the region, such as parallel, task. This patch doesn't include
the Sema checks for that, so the use of them is UB. Simple directives, such as
atomic, can be used. We provide a set of APIs (for C, they are prefix with
`ompx_`; for C++, they are in `ompx` namespace) to get thread id, block id, etc.
For more details, you can refer to
https://tianshilei.me/wp-content/uploads/llvm-hpc-2023.pdf.

jdoerfert commented on clang/lib/CodeGen/CGOpenMPRuntimeGPU.cpp:
```diff
@@ -752,27 +751,41 @@ void CGOpenMPRuntimeGPU::emitNonSPMDKernel(const OMPExecutableDirective &D,
   EntryFunctionState EST;
   WrapperFunctionsMap.clear();
 
+  bool IsBareKernel = D.getSingleClause<OMPXBareClause>();
```
How can a non-spmd kernel be bare? We should force bare kernels to be SPMD, make this an assertion.

shiltian replied:
So `target teams` is generic mode. Since we don't emit the kernel init/deinit, the kernel env will not be emitted as well. At runtime we will default to SPMD mode.
As discussed offline, we will need an extra mode, if we want, but no kernel env should work perfectly fine.

shiltian replied:
I made the change in the parent scope such that a bare kernel is taken as SPMD as well.

jdoerfert reviewed (commented):
Where is the code that prevents the init/deinit calls from being created?

shiltian commented:
> Where is the code that prevents the init/deinit calls from being created?

It's in the early exit.

shiltian committed:
[OpenMP] Introduce the initial support for OpenMP kernel language

This patch starts the support for OpenMP kernel language, basically to write
OpenMP target region in SIMT style, similar to kernel languages such as CUDA.
What included in this first patch is the `ompx_bare` clause for `target teams`
directive. When `ompx_bare` exists, globalization is disabled such that local
variables will not be globalized. The runtime init/deinit function calls will
not be emitted. That being said, almost all OpenMP executable directives are
not supported in the region, such as parallel, task. This patch doesn't include
the Sema checks for that, so the use of them is UB. Simple directives, such as
atomic, can be used. We provide a set of APIs (for C, they are prefix with
`ompx_`; for C++, they are in `ompx` namespace) to get thread id, block id, etc.
For more details, you can refer to
https://tianshilei.me/wp-content/uploads/llvm-hpc-2022.pdf?swcfpc=1.

jdoerfert reviewed (approved):
LG, we need runtime tests in a follow up.

jplehr commented:
It seems that this broke the AMDGPU OpenMP buildbot https://lab.llvm.org/buildbot/#/builders/193/builds/39393
I saw that you have since pushed up one patch regarding pointer compares. Are you looking at the remaining test fails as well?
